### ISSUE TITLE
Unable to get this working in Slim

### ISSUE THREAD

**Issue by rohitsodhia** [tuupola/cors-middleware#27]
I wanted to make sure: is the documentation up to date? I've tried this in various formats, and it doesn't work, and looking at the code, I don't really understand how what's mentioned in the documentation would work, but I'm chalking that up to my not understanding PSR 7 and 15.

I'm on Slim 3.9.2. So far I've tried:

    $app->add(new Tuupola\Middleware\CorsMiddleware);

    $app->add(new Tuupola\Middleware\CorsMiddleware([
		"origin" => [env('FRONTEND_DOMAIN')],
		"methods" => ['*'],
		"headers.allow" => ['*'],
		"headers.expose" => [],
		"credentials" => false,
		"cache" => 0,
	]);

	$app->add(function (Request $request, Response $response, $next) {
		$corsMiddleware = new Tuupola\Middleware\CorsMiddleware([
			"origin" => [env('FRONTEND_DOMAIN')],
			"methods" => $methods,
			"headers.allow" => $request->getHeader('access-control-request-headers'),
			"headers.expose" => [],
			"credentials" => false,
			"cache" => 0,
		]);
		return $corsMiddleware;;
	});

I've also tried the last one returning `->process($request, $response)`.

Nothing seems to add any CORS headers to my calls. I'd love to find out where I'm going wrong.

**Comment by DePalmo** [comment]
Same here, the presented default configuration in readme does not work ... The only way for me right now to return those headers, is to manually include them like this:
```
$app->add(function ($req, $res, $next) {
	$response = $next($req, $res);
	return $response
		->withHeader('Access-Control-Allow-Origin', 'http://mysite')
		->withHeader('Access-Control-Allow-Headers', 'X-Requested-With, Content-Type, Accept, Origin, Authorization')
		->withHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS');
});

```

**Comment by tuupola** [comment]
The default configuration works just fine. 

```php
<?php

require __DIR__ . "/vendor/autoload.php";

$app = new \Slim\App();

$app->add(new Tuupola\Middleware\CorsMiddleware([
    "origin" => ["*"],
    "methods" => ["GET", "POST", "PUT", "PATCH", "DELETE"],
    "headers.allow" => [],
    "headers.expose" => [],
    "credentials" => false,
    "cache" => 0,
]));

$app->get("/", function($request, $response, $arguments) {
    print "here be dragons";
});

$app->run();
```

```
$ curl "http://127.0.0.1:8080" --request OPTIONS --include --header "Origin: http://www.example.com" --header "Access-Control-Request-Method: PUT"

HTTP/1.1 200 OK
Host: 127.0.0.1:8080
Date: Sun, 28 Oct 2018 15:50:22 +0000
Connection: close
X-Powered-By: PHP/7.1.14
Access-Control-Allow-Origin: http://www.example.com
Vary: Origin
Access-Control-Allow-Methods: GET, POST, PUT, PATCH, DELETE
Content-Length: 0
```

@DePalmo Are you sure you are actually making a [CORS request](https://www.html5rocks.com/static/images/cors_server_flowchart.png)? If yes can you paste  an example curl request here. No screenshots from Postman, they are not useful.

 @rohitsodhia Neither of above examples are from the documentation. However I tested and the first one does work.  Note that `*` does not really do anything useful with `methods` nor `headers.allow`. The `*` wildcard is only usable with  `origin` setting.

The second one cannot work at all and Slim will throw `UnexpectedValueException`when a HTTP request is made.

**Comment by DePalmo** [comment]
@tuupola I re-checked and no, didn't. Once I did, I started to see the headers appended. Sorry for the confusion, but I was under impression that those headers should be added to all requests, not only when doing a CORS request. Will see how Angular will accept that ...

**Comment by tuupola** [comment]
Yeah, the headers are added only to some specific requests. The CORS spec is quite confusing as you can see from the flowchart...

**Comment by DePalmo** [comment]
I noticed ... thanks again for clearing some of the confusion.
